crates/ide-assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -246,6 +246,29 @@ fn create_struct_def(
         Either::Left(field_list) => {
             let field_list = field_list.clone_for_update();
 
+            // replace `Self` with the enum name when construct struct def
+            field_list
+                .fields()
+                .filter_map(|field| match field.ty()? {
+                    ast::Type::PathType(p) => Some(
+                        p.syntax()
+                            .descendants_with_tokens()
+                            .filter_map(|it| {
+                                if it.kind() == T![Self] {
+                                    let type_arg =
+                                        make::type_arg(make::ty(&enum_.name()?.to_string()))
+                                            .clone_for_update();
+                                    Some(ted::replace(it, type_arg.syntax()))
+                                } else {
+                                    None
+                                }
+                            })
+                            .count(),
+                    ),
+                    _ => None,
+                })
+                .count();
```

Comment:
This should use the `PathTransform::transform_path` function to transform the `Self` path